### PR TITLE
drag event static helper methods + unit tests

### PR DESCRIPTION
**Pull request by afshin**
These are three static methods to help subclasses implement drag and drop functionality on top of phosphor-widget. They abstract away the fact that we care about a specific MIME type: `application/x-phosphor-widget`, but this MIME type may need to be exposed at some point in the future.

One of the first use cases will be for implementing a dashboard panel with draggable/droppable widgets.

### CODE REVIEW

**Comment by dwillmer** [comment]
LGTM, +1

File: src/index.ts [hunk 1]
```diff
@@ -141,6 +141,18 @@ export
...
+ * The mime type for Widget drag and drop operations, for use with the native
+ * DataTransfer object's setData and getData methods.
+ */
+const WIDGET_MIME_TYPE = 'application/x-phosphor-widget';
+
+/**
+ * The flag value in dragging events, for use with the native DataTransfer's
+ * setData and getData methods. Any non-empty string will work.
+ */
+const WIDGET_MIME_DATA = 'dummy';
```
Comment (sccolbert) [inline src/index.ts]: Let's make this something unique (but human readable), and then check for for equality in `getDragMimeData`.

File: src/index.ts [hunk 2]
```diff
@@ -141,6 +141,18 @@ export
 const MSG_BEFORE_DETACH = new Message('before-detach');
 
 /**
+ * The mime type for Widget drag and drop operations, for use with the native
+ * DataTransfer object's setData and getData methods.
+ */
+const WIDGET_MIME_TYPE = 'application/x-phosphor-widget';
```
Comment (sccolbert) [inline src/index.ts]: would `x-phosphor-widget-factory` be more descriptive?

File: src/index.ts (hunk 1 above)
Comment (afshin) [inline src/index.ts]: Sure. Will do.

File: src/index.ts [hunk 3]
```diff
@@ -150,6 +162,7 @@ const WIDGET_CLASS = 'p-Widget';
  */
 const HIDDEN_CLASS = 'p-mod-hidden';
 
+let dragFactory: () => Widget = null;
```
Comment (sccolbert) [inline src/index.ts]: I like to put public and semantically relevant things at the top of the file. An non-relevant, private things at the bottoms. This means someone reading the file sees everything they need at the top. 

So, let's make this a `var` near the bottom of the file with the rest of the module-private stuff.

File: src/index.ts (hunk 2 above)
Comment (afshin) [inline src/index.ts]: Yes, it would. I was just thinking that `setDragMimeData` *only* allows setting factories.

File: src/index.ts [hunk 4]
```diff
@@ -268,6 +281,35 @@ class Widget extends NodeWrapper implements IDisposable, IMessageHandler {
   });
 
   /**
+   * Clear the internal reference to a DragEvent's eventual handler, which
+   * is stored in the local private variable: dragFactory. This method should
+   * be called in dragend event handlers.
+   */
```
Comment (sccolbert) [inline src/index.ts]: Our docs format is one-line summary follow by blank line. Then additional description (if needed). Followed by args/return spec, then notes, example, and see-also refs. Here's an example to follow:

https://github.com/phosphorjs/phosphor-arrays/blob/master/src/index.ts#L11

Sometimes it's not possible to condense the description into a single line, and that's fine. But the goal should be a single line.

This line note applies to all docs added by this PR.

File: src/index.ts (hunk 3 above)
Comment (afshin) [inline src/index.ts]: Right, that `var` vs `let` issue you pointed out. Cool.

File: src/index.ts [hunk 5]
```diff
@@ -268,6 +281,35 @@ class Widget extends NodeWrapper implements IDisposable, IMessageHandler {
...
+    return event.dataTransfer.getData(WIDGET_MIME_TYPE) ? dragFactory : null;
+  }
+
+  /**
+   * Set the internal reference to a DragEvent's eventual handler, which is
+   * store in the local private variable: dragFactory. This method should be
+   * called in dragstart event handlers.
+   */
+  static setDragMimeData(event: DragEvent, factory: () => Widget): void {
+    if (factory) {
```
Comment (sccolbert) [inline src/index.ts]: Assume that factory won't be `null` or `undefined`.

File: src/index.ts [hunk 6]
```diff
@@ -268,6 +281,35 @@ class Widget extends NodeWrapper implements IDisposable, IMessageHandler {
...
+      dragFactory = factory;
+      event.dataTransfer.setData(WIDGET_MIME_TYPE, WIDGET_MIME_DATA);
```
Comment (sccolbert) [inline src/index.ts]: actually, now that i think about it, the `WIDGET_MIME_DATA` should be a id which is created anew each time this method is called. The `getDragMimeData` should check that id before returning the factory. This ensures that the correct factory is always associated with the correct event sequence.

File: src/index.ts (hunk 4 above)
Comment (afshin) [inline src/index.ts]: Noted, will update.

File: src/index.ts [hunk 7]
```diff
@@ -1426,7 +1468,6 @@ class ChangeTitleMessage extends Message {
   private _title: string;
 }
 
-
```
Comment (sccolbert) [inline src/index.ts]: This blank line should not be removed.

File: test/src/index.ts [hunk 8]
```diff
@@ -30,7 +30,6 @@ import {
 
 import './index.css';
 
-
```
Comment (sccolbert) [inline test/src/index.ts]: This blank line should not be removed. We use 2 blanks to separate module-level entities.

File: src/index.ts (hunk 5 above)
Comment (afshin) [inline src/index.ts]: :heavy_check_mark:

File: src/index.ts (hunk 6 above)
Comment (afshin) [inline src/index.ts]: But there's no notion of multiple events, drags are by their nature totally atomic and synchronous. You can't get another drag's data.

Comment (sccolbert) [inline src/index.ts]: But these methods could be called at any time between the start of drag and the drop. If that happens, the original intended factory would be gone, and the drop target would get the wrong thing. It's better to get nothing, than the wrong thing. This is just a defensive measure to protect against wonky user code.

Comment (afshin) [inline src/index.ts]: Good point.

**Comment by sccolbert** [comment]
This is being moved to a PR against `phosphor-domutil` with more generality.

**Pull request closed without merging**